src/connector/src/source/cdc/external/sql_server.rs

```diff
@@ -0,0 +1,467 @@
+use futures_async_stream::try_stream;
+use itertools::Itertools;
+use risingwave_common::bail;
+use risingwave_common::catalog::{ColumnDesc, ColumnId, Schema};
+use risingwave_common::row::OwnedRow;
+use risingwave_common::types::{DataType, ScalarImpl};
+use serde_derive::{Deserialize, Serialize};
+use tiberius::error::Error;
+use tiberius::{ColumnType, Config, Query, QueryItem};
+use tokio::net::TcpStream;
+use tokio_util::compat::{Compat, TokioAsyncWriteCompatExt};
+
+use crate::error::{ConnectorError, ConnectorResult};
+use crate::parser::{sql_server_row_to_owned_row, ScalarImplTiberiusWrapper};
+use crate::sink::sqlserver::SqlServerClient;
+use crate::source::cdc::external::{
+    CdcOffset, CdcOffsetParseFunc, DebeziumOffset, ExternalTableConfig, ExternalTableReader,
+    SchemaTableName,
+};
+
+// The maximum commit_lsn value in Sql Server
+const MAX_COMMIT_LSN: &str = "ffffffff:ffffffff:ffff";
+
+#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
+pub struct SqlServerOffset {
+    // https://learn.microsoft.com/en-us/answers/questions/1328359/how-to-accurately-sequence-change-data-capture-dat
+    pub change_lsn: String,
+    pub commit_lsn: String,
+}
+
+// only compare the lsn field
+impl PartialOrd for SqlServerOffset {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        match self.change_lsn.partial_cmp(&other.change_lsn) {
+            Some(Ordering::Equal) => self.commit_lsn.partial_cmp(&other.commit_lsn),
+            other => other,
+        }
+    }
+}
+
+impl SqlServerOffset {
+    pub fn parse_debezium_offset(offset: &str) -> ConnectorResult<Self> {
+        let dbz_offset: DebeziumOffset = serde_json::from_str(offset)
+            .with_context(|| format!("invalid upstream offset: {}", offset))?;
+
+        Ok(Self {
+            change_lsn: dbz_offset
+                .source_offset
+                .change_lsn
+                .context("invalid sql server change_lsn")?,
+            commit_lsn: dbz_offset
+                .source_offset
+                .commit_lsn
+                .context("invalid sql server commit_lsn")?,
+        })
+    }
+}
+
+pub struct SqlServerExternalTable {
+    column_descs: Vec<ColumnDesc>,
+    pk_names: Vec<String>,
+}
+
+impl SqlServerExternalTable {
+    pub async fn connect(config: ExternalTableConfig) -> ConnectorResult<Self> {
+        tracing::debug!("connect to sql server");
+
+        let mut client_config = Config::new();
+
+        client_config.host(&config.host);
+        client_config.database(&config.database);
+        client_config.port(config.port.parse::<u16>().unwrap());
+        client_config.authentication(tiberius::AuthMethod::sql_server(
+            &config.username,
+            &config.password,
+        ));
+        client_config.trust_cert();
+
+        let tcp = TcpStream::connect(client_config.get_addr()).await?;
+        tcp.set_nodelay(true)?;
+
+        let mut client: tiberius::Client<Compat<TcpStream>> =
```

Comment:
It seems didn't reuse the `SqlServerClient` before?

Reply:
updated